File: Games/maths/MathsFun/MathsFun/cppLib/geometry2d.cpp
#include "geometry2d.h"
#include "matrices.h"
#include <cmath>
#include <cfloat>

// combination of absolute and relative tolerances.
// c.f. https://www.realtimecollisiondetection.net/pubs/Tolerances/
#define CMP(x, y)                               \
    (fabsf((x)-(y)) <= FLT_EPSILON *            \
        fmaxf(1.0f,                             \
              fmaxf(fabsf(x), fabsf(y))))

float Length(const Line2D &line) {
    return Magnitude(line.end - line.start);
} // Length

float LengthSq(const Line2D &line) {
    return MagnitudeSq(line.end - line.start);
} // LengthSq

vec2 GetMin(const Rectangle2D &rect) {
    vec2 p1 = rect.origin;
    vec2 p2 = rect.origin + rect.size;
    return vec2(fminf(p1.x, p2.x), fminf(p1.y, p2.y));
} // GetMin

vec2 GetMax(const Rectangle2D &rect) {
    vec2 p1 = rect.origin;
    vec2 p2 = rect.origin + rect.size;
    return vec2(fmaxf(p1.x, p2.x), fmaxf(p1.y, p2.y));
} // GetMax

Rectangle2D FromMinMax(const vec2 &min, const vec2 &max) {
    return Rectangle2D(min, max - min);
} // FromMinMax

bool PointOnLine(const Point2D &point, const Line2D &line) {
    float dx = line.end.x - line.start.x;
    float dy = line.end.y - line.start.y;
    float m = dy / dx;
    float b = line.start.y - m * line.start.x;
    return CMP(point.y, m * point.x + b);
} // PointOnLine

bool PointInCircle(const Point2D &point, const Circle &circle) {
    Line2D line(point, circle.position);
    if (LengthSq(line) < circle.radius * circle.radius) {
        return true;
    }
    return false;
} // PointInCircle

bool PointInRectangle(const Point2D &point, const Rectangle2D &rectangle) {
    vec2 min = GetMin(rectangle);
    vec2 max = GetMax(rectangle);

    return min.x <= point.x
        && min.y <= point.y
        && point.x <= max.x
        && point.y <= max.y;
    return false;
} // PointInRectangle

bool PointInOrientedRectangle(const Point2D &point, const OrientedRectangle &rectangle) {
    vec2 rotVector = point - rectangle.position;
    float theta = -DEG2RAD(rectangle.rotation);
    float zRotation2x2[] = {
        cosf(theta), sinf(theta),
        -sinf(theta), cosf(theta)
    };
    Multiply(rotVector.asArray,
             vec2(rotVector.x, rotVector.y).asArray, 1, 2,  // array/rows/cols
             zRotation2x2, 2, 2);
    Rectangle2D localRectangle(Point2D(), rectangle.halfExtents * 2.0f);
    vec2 localPoint = rotVector + rectangle.halfExtents;
    return PointInRectangle(localPoint, localRectangle);
} // PointInOrientedRectangle

bool LineCircle(const Line2D &line, const Circle &circle) {
    vec2 ab = line.end - line.start;
    float t = Dot(circle.position - line.start, ab) / Dot(ab, ab);
    if (t < 0.0f || t > 1.0f) {
        return false;
    }
    Point2D closestPoint = line.start + ab * t;
    Line2D circleToClosest(circle.position, closestPoint);
    return LengthSq(circleToClosest) < circle.radius * circle.radius;
} // LineCircle

bool LineRectangle(const Line2D &line, const Rectangle2D &rectangle) {
    if (PointInRectangle(line.start, rectangle) || PointInRectangle(line.end, rectangle)) {
        return true;
    }

    vec2 norm = Normalized(line.end - line.start);
    norm.x = (norm.x != 0) ? 1.0f / norm.x : 0;
    norm.y = (norm.y != 0) ? 1.0f / norm.x : 0;

    vec2 min = (GetMin(rectangle) - line.start) * norm;
    vec2 max = (GetMax(rectangle) - line.start) * norm;

    float tmin = fmaxf(
        fminf(min.x, max.x), fminf(min.y, max.y));
    float tmax = fminf(
        fmaxf(min.x, max.x), fmaxf(min.y, max.y));
    if (tmax < 0 || tmin > tmax) {
        return false;
    }

    float t = (tmin < 0.0f) ? tmax : tmin;
    return t > 0.0f && t * t < LengthSq(line);
} // LineRectangle

bool LineOrientedRectangle(const Line2D &line, const OrientedRectangle &rectangle) {
    float theta = -DEG2RAD(rectangle.rotation);
    float zRotation2x2[] = {
        cosf(theta), sinf(theta),
        -sinf(theta), cosf(theta)
    };
    Line2D localLine;

    vec2 rotVector = line.start - rectangle.position;
    Multiply(rotVector.asArray,
             vec2(rotVector.x, rotVector.y).asArray,
             1, 2, zRotation2x2, 2, 2);
    localLine.start = rotVector + rectangle.halfExtents;

    rotVector = line.end - rectangle.position;
    Multiply(rotVector.asArray,
             vec2(rotVector.x, rotVector.y).asArray,
             1, 2, zRotation2x2, 2, 2);
    localLine.end = rotVector + rectangle.halfExtents;

    Rectangle2D localRectangle(Point2D(), rectangle.halfExtents * 2.0f);
    return LineRectangle(localLine, localRectangle);

} // LineOrientedRectangle

bool CircleCircle(const Circle &c1, const Circle &c2) {
    Line2D line(c1.position, c2.position);
    float radiiSum = c1.radius + c2.radius;

    return LengthSq(line) <= radiiSum * radiiSum;
} // CircleCircle


bool CircleRectangle(const Circle &circle, const Rectangle2D &rectangle) {
    vec2 min = GetMin(rectangle);
    vec2 max = GetMax(rectangle);

    // find closest point on the rectangle
    Point2D closestPoint = circle.position;
    if (closestPoint.x < min.x) {
        closestPoint.x = min.x;
    }
    else if (closestPoint.x > max.x) {
        closestPoint.x = max.x;
    }

    if (closestPoint.y < min.y) {
        closestPoint.y = min.y;
    }
    else if (closestPoint.y > max.y) {
        closestPoint.y = max.y;
    }

    Line2D line(circle.position, closestPoint);
    return LengthSq(line) <= circle.radius * circle.radius;

} // CircleRectangle

bool CircleOrientedRectangle(const Circle &circle, const OrientedRectangle &rectangle) {
    vec2 r = circle.position - rectangle.position;
    float theta = -DEG2RAD(rectangle.rotation);
    float zRotation2x2[] = {
        cosf(theta), sinf(theta),
        -sinf(theta), cosf(theta)
    };

    Multiply(r.asArray, vec2(r.x, r.y).asArray, 1, 2,
             zRotation2x2, 2, 2);
    Circle localCircle(r + rectangle.halfExtents, circle.radius);
    Rectangle2D localRect(Point2D(), rectangle.halfExtents * 2.0f);

    return CircleRectangle(localCircle, localRect);

} // CircleOrientedRectangle

bool RectangleRectangle(const Rectangle2D &rect1, const Rectangle2D &rect2) {
    vec2 aMin = GetMin(rect1);
    vec2 aMax = GetMax(rect1);

    vec2 bMin = GetMin(rect2);
    vec2 bMax = GetMax(rect2);

    bool overX = ((bMin.x <= aMax.x) && (aMin.x <= bMax.x));
    bool overY = ((bMin.y <= aMax.y) && (aMin.y <= bMax.y));

    return overX && overY;

} // RctangleRectangle

Interval2D GetInterval(const Rectangle2D &rect, const vec2 &axis) {
    Interval2D result;

    vec2 min = GetMin(rect);
    vec2 max = GetMax(rect);

    vec2 verts[] = {
        vec2(min.x, min.y), vec2(min.x, max.y),
        vec2(max.x, max.y), vec2(max.x, min.y),
    };

    // project each vertex onto the axis, store the smallest and largest values
    result.min = result.max = Dot(axis, verts[0]);

    for (int i = 1; i < 4; i++) {
        float projection = Dot(axis, verts[i]);
        if (projection < result.min) result.min = projection;
        if (projection > result.max) result.max = projection;
    }

    return result;
} // GetInterval

bool OverlapOnAxis(const Rectangle2D &rect1, const Rectangle2D &rect2,
                   const vec2 &axis) {
    Interval2D a = GetInterval(rect1, axis);
    Interval2D b = GetInterval(rect2, axis);
    return ((b.min <= a.max) && (a.min <= b.max));
} // OverlapOnAxis

bool RectangleRectangleSAT(const Rectangle2D &rect1, const Rectangle2D &rect2) {
    vec2 axisToTest[] = { vec2(1, 0), vec2(0, 1) };
    
    for (int i = 0; i < 2; i++) {
        if (!OverlapOnAxis(rect1, rect2, axisToTest[i])) {
            return false;
        }
    }
    return true;
} // RectangleRectangleSAT

File: Games/maths/MathsFun/MathsFun/swiftLib/Geometry2D.swift
import Foundation

typealias Point2D = Vec2

struct Line2D: Equatable {
    let start: Point2D
    let end: Point2D

    init(start: Point2D, end: Point2D) {
        self.start = start
        self.end = end
    }

    init(_ contents: Double...) {
        self.start = Point2D(x: contents[0], y: contents[1])
        self.end = Point2D(x: contents[2], y: contents[3])
    }

    var length: Double {
        (end - start).magnitude
    }

    var lengthSquared: Double {
        (end - start).magnitudeSquared
    }

    func contains(_ point: Point2D?) -> Bool {
        guard let point else { return false }

        let dx = end.x - start.x
        let dy = end.y - start.y
        let m = dy / dx
        let b = start.y - m * start.x
        return absRelFPCompare(point.y, m * point.x + b)
    }
}

struct Circle: Equatable {
    var position: Point2D
    var radius: Double

    init(position: Point2D, radius: Double) {
        self.position = position
        self.radius = radius
    }

    init(_ contents: Double...) {
        self.position = Point2D(x: contents[0], y: contents[1])
        self.radius = contents[2]
    }

    func contains(_ point: Point2D?) -> Bool {
        guard let point else { return false }

        let line = Line2D(start: point, end: position)
        if line.lengthSquared < radius * radius {
            return true
        }
        return false
    }

    func intersects(_ line: Line2D?) -> Bool {
        guard let line else { return false }

        let ab = line.end - line.start
        let t = (position - line.start).dot(ab) / ab.dot(ab)
        if t < 0.0 || t > 1.0 {
            return false
        }

        let closestPoint = line.start + ab * t
        let circleToClosest = Line2D(start: position, end: closestPoint)
        return circleToClosest.lengthSquared < radius * radius
    }

    func intersects(_ c2: Circle) -> Bool {
        let line = Line2D(start: position, end: c2.position)
        let radiiSum = radius + c2.radius
        return line.lengthSquared <= radiiSum * radiiSum
    }
    
    func intersects(_ rect: Rectangle2D) -> Bool {
        var closestPoint = position
        let min = rect.min
        let max = rect.max

        // a clamping we will go...
        if closestPoint.x < min.x {
            closestPoint.x = min.x
        }
        else if closestPoint.x > max.x {
            closestPoint.x = max.x
        }

        if closestPoint.y < min.y {
            closestPoint.y = min.y
        }
        else if closestPoint.y > max.y {
            closestPoint.y = max.y
        }

        let line = Line2D(start: position, end: closestPoint)
        return line.lengthSquared <= radius * radius
        
    }
}

extension Line2D {
    func intersects(_ circle: Circle?) -> Bool {
        guard let circle else { return false }
        return circle.intersects(self)
    }
}

struct Rectangle2D: Equatable {
    var origin: Point2D
    var size: Vec2

    init(origin: Point2D, size: Vec2) {
        self.origin = origin
        self.size = size
    }

    init(_ contents: Double...) {
        self.origin = Point2D(x: contents[0], y: contents[1])
        self.size = Vec2(x: contents[2], y: contents[3])
    }

    static func fromMinMax(min: Vec2, max: Vec2) -> Rectangle2D {
        Rectangle2D(origin: min, size: max - min)
    }

    var min: Vec2 {
        let p1 = origin
        let p2 = origin + size
        return Vec2(x: fmin(p1.x, p2.x), y: fmin(p1.y, p2.y))
    }

    var max: Vec2 {
        let p1 = origin
        let p2 = origin + size
        return Vec2(x: fmax(p1.x, p2.x), y: fmax(p1.y, p2.y))
    }

    func contains(_ point: Point2D?) -> Bool {
        guard let point else { return false }

        let min = self.min
        let max = self.max

        return min.x <= point.x
          && min.y <= point.y
          && point.x <= max.x
          && point.y <= max.y
    }

    func intersects(_ line: Line2D?) -> Bool {
        guard let line else { return false }
        if contains(line.start) || contains(line.end) { return true }
        
        var norm = (line.end - line.start).normalized
        norm.x = (norm.x != 0) ? 1.0 / norm.x : 0
        norm.y = (norm.y != 0) ? 1.0 / norm.y : 0

        let minv = (min - line.start) * norm
        let maxv = (max - line.start) * norm

        let tmin = fmax(
          fmin(minv.x, maxv.x),
          fmin(minv.y, maxv.y))
        let tmax = fmin(
          fmax(minv.x, maxv.x),
          fmax(minv.y, maxv.y))

        if tmax < 0 || tmin > tmax {
            return false
        }

        let t = (tmin < 0.0) ? tmax : tmin
        return t > 0 && t*t < line.lengthSquared
    }

    func intersects(_ circle: Circle) -> Bool {
        circle.intersects(self)
    }
}

extension Line2D {
    func intersects(_ rectangle: Rectangle2D?) -> Bool {
        guard let rectangle else { return false }
        return rectangle.intersects(self)
    }
}

struct OrientedRectangle: Equatable {
    let position: Point2D
    let halfExtents: Vec2
    let rotationDegrees: Double

    init() {
        self.position = Point2D()
        self.halfExtents = Vec2(x: 1.0, y: 1.0)
        self.rotationDegrees = 0.0
    }

    init(position: Point2D, halfExtents: Vec2) {
        self.position = position
        self.halfExtents = halfExtents
        self.rotationDegrees = 0.0
    }

    init(position: Point2D, halfExtents: Vec2, rotation: Double) {
        self.position = position
        self.halfExtents = halfExtents
        self.rotationDegrees = rotation
    }

    func contains(_ point: Point2D?) -> Bool {
        guard let point else { return false }

        let theta = -rotationDegrees.degreesToRadians
        let zRotation2x2 = Mat2(
          cos(theta), sin(theta),
          -sin(theta), cos(theta))

        let rotVector = (point - position) * zRotation2x2
        
        let localRectangle = Rectangle2D(origin: Point2D(),
                                         size: halfExtents * 2.0)
        let localPoint = rotVector + halfExtents
        return localRectangle.contains(localPoint)
    }

    func intersects(_ line: Line2D?) -> Bool {
        guard let line else { return false }
        
        let theta = -rotationDegrees.degreesToRadians
        let zRotation2x2 = Mat2(
          cos(theta), sin(theta),
          -sin(theta), cos(theta))

        let rotVector1 = (line.start - position) * zRotation2x2
        let rotVector2 = (line.end - position) * zRotation2x2
        
        let localLine = Line2D(start: rotVector1 + halfExtents,
                               end: rotVector2 + halfExtents)
        let localRectangle = Rectangle2D(origin: Point2D(), size:  halfExtents * 2.0)

        return localRectangle.intersects(localLine)
    }
}

extension Line2D {
    func intersects(_ orectangle: OrientedRectangle?) -> Bool {
        guard let orectangle else { return false }
        return orectangle.intersects(self)
    }
}


